discourse: Updates and fixes

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Add the stuff that was missing in #127931, completing the plugin system update:

- Add an `update-plugins` action to the updater
- Update the documentation to reflect the plugin system changes
- Upgrade plugin versions and add missing metadata
- Provide a package with all plugins enabled, useful to avoid local builds when all plugins are wanted

Also:
- Make the updater remove the architecture bound platforms from `Gemfile.lock` and `libv8` architecture bound entries - without this, discourse fails to start with the latest bundler version
- Update Discourse to version `2.7.5`
- Use Terser when building the assets

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

talyz committed:
discourse.plugins: Update and add metadata

talyz committed:
discourse: Add `update-plugins` action to update.py

talyz committed:
discourseAllPlugins: Provide a discourse derivation with all plugins

talyz committed:
nixos/discourse: Update plugin documentation

Update the documentation regarding plugins to reflect recent changes.

ryantm reviewed (approved):
Looks good

lukegb commented on pkgs/servers/web-apps/discourse/plugins/discourse-canned-replies/default.nix:
```diff
@@ -1,11 +1,17 @@
+  meta = with lib; {
+    homepage = "https://github.com/discourse/discourse-canned-replies";
+    maintainers = with maintainers; [ talyz ];
+    license = licenses.gpl2;
```
gpl2Only: https://github.com/discourse/discourse-canned-replies/blob/main/package.json#L3

talyz replied:
Oh, right! Thanks!

lukegb reviewed (approved):
LGTM, one nit

SuperSandro2000 commented on pkgs/servers/web-apps/discourse/update.py:
```diff
@@ -35,6 +38,15 @@ def tags(self) -> Iterable[str]:
+    @property
+    def latest_commit_sha(self) -> str:
+        if self._latest_commit_sha is None:
+            r = requests.get(f'https://api.github.com/repos/{self.owner}/{self.repo}/commits?per_page=1')
```
Does this use a token?

talyz replied:
No, you can make these API requests without one. Having one probably lifts the rate limit, but it's at 60 requests without a token, which is plenty.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 129464` at 39dd141e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/f86899ec9254e806195883d8c9d28df5">discourseAllPlugins</a></li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

bundler is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/servers/web-apps/discourse/default.nix:217:5:
```
    |
217 |     buildInputs = [
    |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `replace`.
Near pkgs/servers/web-apps/discourse/default.nix:4:9:
```
  |
4 | , ruby, replace, gzip, gnutar, git, cacert, util-linux, gawk
  |         ^
```
</li>
    <li>warning: unused-argument

Unused argument: `meta`.
Near pkgs/servers/web-apps/discourse/default.nix:56:7:
```
   |
56 |     , meta ? null
   |       ^
```
</li>
    <li>warning: unused-argument

Unused argument: `src`.
Near pkgs/servers/web-apps/discourse/default.nix:58:7:
```
   |
58 |     , src
   |       ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

talyz committed:
discourse: Remove architecture bound platforms from Gemfile.lock

Maybe bundix doesn't handle them properly? They cause runtime issues
and don't seem necessary when the binary gems are built from scratch
anyway.

talyz commented:
Rebased on `master` and pushed another update which removes the architecture bound `libv8` `Gemfile.lock` entries, since they caused runtime errors with the latest version of bundler. The update also makes sure these architecture bound entries are always removed from the lock file by removing the platforms in the updater script.

talyz commented:
I previously forgot about handling the plugin ruby dependencies in the updater, but that's sorted now. I also expanded the manual section to include instructions on how to package plugins with ruby deps.

talyz committed:
discourse: Use `terser` when building assets

Discourse prefers to use `terser` when building js assets, see
https://github.com/discourse/discourse/pull/12656. It still wants to
find `uglify-js` in order to not fall back to a ruby js compression
library, so let's keep it around for now. A fix for this has been
submitted upstream in
https://github.com/discourse/discourse/pull/13683.

dpausp commented on pkgs/servers/web-apps/discourse/update.py:
```diff
@@ -8,17 +8,22 @@
 import re
 import logging
 import subprocess
-import pathlib
```
This breaks `update`, line 169 also needs `pathlib.Path` -> `Path`

talyz replied:
Oh, whoops. Fixed it and added the latest update to `2.7.5`.

dpausp reviewed (changes requested):
`update-plugins` works as expected, there are some new plugin versions. `update` is broken.

talyz committed:
discourse: 2.7.4 -> 2.7.5

dpausp reviewed (approved):
Looks good now. 

Tested all `update.py` commands and also also played around a bit with Discourse itself: tested posting, looked at the admin dashboard and plugin list, and ran a backup (on 21.05 without the terser commit).

dpausp commented:
As a side note, if someone wants to try the beta version (tests-passed), `update.py update v2.8.0.beta2` works and Discourse is running properly.
